### ISSUE TITLE
Give option to merge default config

### ISSUE THREAD

**Issue by atulgpt** [detekt/detekt#6053]
In `detekt` for multiple rules, we set default configs that we think are best for the community. Although some user wants to change the default, the only applicable way is to overwrite the default and copy all the pre-existing default values in the config.

There could be a few ways to solve this:
1. When in the new release if we introduce new values to default configs, we can put some type of update markers in the `detekt.yml` which the user has to resolve(either choose the new values or discard them) and then remove the update marker. With the update marker present in `detekt.yml` any type of `detekt` build will fail.
2. Add the ability to extend from the default `detekt` config and then allow the options to overwrite the configs. This is similar to how Eslint maintains its config and gives the option to overwrite. Although in the case of extending array configs, we might have to introduce some special syntax to provide options like completely overwriting the defaults or extending to it or subtracting some
3. While generating the `detekt.yml` config we can create some type of special keys like `detekt_defaults` which we can append instead of writing the whole list. This `detekt_defaults` can have different values based on the `detekt` version and will automatically pick the newly added ones from the latest release

This issue has also been discussed https://github.com/detekt/detekt/issues/5896#issuecomment-1523218452 and https://github.com/detekt/detekt/issues/5977#issuecomment-1524237122

**Comment by cortinico** [comment]
How is this different than `buildUponDefaultConfig`?

**Comment by atulgpt** [comment]
Hi @cortinico, does `buildUponDefaultConfig` handle pre-existing arrays of detekt defaults? 

For example, if the user wants to add `BigDecimal` in `ForbiddenMethodCall` in addition to detekt defaults, will adding a new value in the `detekt` default(in some later `detekt` release) automatically append both, detekt default values and user-added ones? Or do user have to rebuild the config?

**Comment by cortinico** [comment]
> does `buildUponDefaultConfig` handle pre-existing arrays of detekt defaults?

Nope, the user will have to provide the new array which will override the existing one

**Comment by atulgpt** [comment]
> Nope, the user will have to provide the new array which will override the existing one

Then that will be problematic as the user might not be aware that in `detekt` we have added new value to the default array config

**Comment by cortinico** [comment]
> Then that will be problematic as the user might not be aware that in `detekt` we have added new value to the default array config

But that's the same for any other key. If the user overrides a value, they won't know if the default has changed (array, string, or any other type).

**Comment by atulgpt** [comment]
Yes if a user really intends to overwrite it completely then it's fine. This issue talks about the case when the user wants to merge or intersect his/her own config with the detekt default one. For example, I really trust `detekt` to provide well-researched meaningful defaults of `ForbiddenMethodCall` `methods` config but in addition to `detekt` default I want to add a few more which is related to my project. For those cases, the current way of overwriting `methods` fails when in the new release `detekt` add more useful functions

@BraisGabin also raised the same concern at this https://github.com/detekt/detekt/issues/5977#issuecomment-1504197136

**Comment by cortinico** [comment]
> For those cases, the current way of overwriting `methods` fails when in the new release `detekt` add more useful functions

I understand this scenario. I believe this can be achieved by https://github.com/detekt/detekt/discussions/6035#discussioncomment-5702491

**Comment by atulgpt** [comment]
> > For those cases, the current way of overwriting `methods` fails when in the new release `detekt` add more useful functions
> 
> I understand this scenario. I believe this can be achieved by [#6035 (comment)](https://github.com/detekt/detekt/discussions/6035#discussioncomment-5702491)

I think we will still be needing a way to extend the default configuration. Having the same rule multiple times with different rules doesn't address the issue mentioned in this issue. Do you have some proposals on how it might work?

**Comment by cortinico** [comment]
> Do you have some proposals on how it might work?

In a sense, if you want to extend the config of a rule, rather than overriding it you should:
* Enable the original rule provided by the Detekt.
* Instantiate the same rule twice with your custom extension only.

**Comment by atulgpt** [comment]
> 

Yes, that would work. But TBH quite unusual use of applying the same rule twice. Need to check how easy it would be for users to configure

**Comment by cortinico** [comment]
> Yes, that would work. But TBH quite unusual use-case of applying the same rule twice. Need to check how easy it would be for users to configure

I would refrain from adding yet another extension point if we're already going to develop one though.
